features/storage/system_storage/SystemStorage.cpp

```diff
@@ -0,0 +1,62 @@
+#include "SystemStorage.h"
+
+namespace mbed {
+namespace SystemStorage{
```

Comment:
Open for discussion, but I don't think we need a new namespace. @deepikabhavnani is working on recovering the mbed namespace over here: https://github.com/ARMmbed/mbed-os/pull/7760

If you look at other libraries, a single namespace is sufficient and prevents users from having to search for names:
- standard C for one: namespace std
- [OpenCL](https://www.khronos.org/opencl/): namespace cl
- [Clang](https://clang.llvm.org/doxygen/classclang_1_1NamespaceDecl.html): namespace clang
- [ROS](http://docs.ros.org/melodic/api/roscpp/html/): namespace ros
- [Ogre3d](https://ogrecave.github.io/ogre/api/latest/index.html): namespace Ogre

IMO `mbed::get_default_blockdevice()` is sufficient

Reply:
The design based on namespace was lead by @kjbracey-arm as architect and we agreed with his proposal

Reply:
Having said that, I'm not totally convinced of the need for the `SystemStorage` wrapping either - the namespace was a counter-proposal to the previous draft which had `SystemStorage` as an object so you needed to do `System_Storage::get_instance().get_default_filesystem()`. The namespace simplified that.

Reply:
Oh I didn't realize Kernel was a namespace. Out of curiousity could it not have been a class with only static member functions?

It's surprising to me to see a namespace used this way but don't see anything inherently wrong with it.

This raises two nitpicking questions:
1. Why not just "Storage", all functions are inherently a part of the "System", I'm not sure the "System prefix adds information.
2. Should the be lower case to match naming conventions?
3. Should we adopt a similar namespace thing for the global event queues?

Reply:
I would vote for 1 namespace to keep things simple.

Reply:
We need a standard for global functions. Currently:
- `mbed::core_util_critical_section_enable()`
- `mbed::Kernel::get_ms_count()`
- `mbed::mbed_event_queue()`

My vote is lets just go with `mbed::get_default_blockdevice()`. Do we expect name conflicts here?

@yossi2le, Sorry this discussion is occuring on your PR.

Reply:
I am not sure why we need special class for get_default_blockdevice? We can adapt the netowork model, where each class gives the default. https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/EthInterface.h#L46

Which will help in getting default from target as well. https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/EthInterface.h#L58

And selection of default based on conf is here https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/NetworkInterfaceDefaults.cpp

Reply:
@deepikabhavnani we followed the pattern that you suggested, but @kjbracey-arm pointed correctly that because we have no data, we do not need a class and better use namespace.
The advantage of using system_storage (now in lower case) as namespace is in preparation for features in storage roadmap that will allow mapping of different storage solutions to different usages at later integration stage instead of directly choosing them on usage.

Reply:
> Oh I didn't realize Kernel was a namespace. Out of curiousity could it not have been a class with only static member functions?

It could. But more recent APIs (eg in C++11 standard library) seem to be adopting the model of using a namespace for something that is backed by an "implicit singleton" but no object of such a class could never actually be instantiated. `this_thread` is a good example.

> We need a standard for global functions.
>  *  mbed::core_util_critical_section_enable()
> *   mbed::Kernel::get_ms_count()
> *   mbed::mbed_event_queue()

The `Kernel` one was modelled to fit alongside all the other `namespace rtos` functions which are `Mutex::xxx` etc (and it was planned to be static member of class `Kernel` originally). There are no C-like functions inside `rtos`, so there's consistency inside that namespace.

For this particular instantation model, we already have `NetworkInterface::get_default_instance()`, `WiFiInterface::get_default_instance()`, `EMAC::get_default_instance()`, `OnboardNetworkStack::get_default_instance()` and more.

This PR does deviate from that pattern for default instance generation, and I'm not totally sure why it deviates. I would favour `FileSystem::get_default_instance()` and `BlockDevice::get_default_instance()`. Especially as we have no other current members of `SystemStorage`.

`mbed_event_queue()` predates the above instance generation - it would be nice to make it follow the above pattern as `EventQueue::get_default_instance()` and `EventQueue:::get_highprio_instance()`.

Looking at the all the above, it seems like the style is not to create namespaces inside `mbed`, at least not lightly. I remain open to `SystemStorage`, but at the minute it just has the 2 generators I think would be better in the separate classes, and it feels a bit broad, potentially encompassing all sorts of software bits that fit at different layers.